### PR TITLE
Expose per zone hosts as part of host set

### PR DESCRIPTION
**Pull request by RomanDzhabarov**
This will be required for advanced load balancing when consider local cluster (This is going to be separate change).

At this point I'd like to get proof of concept for routing and then we can optimize it further (e.g., precompute % of the healthy zone hosts to all hosts).

### CODE REVIEW

File: test/common/upstream/logical_dns_cluster_test.cc [hunk 1]
```diff
@@ -78,8 +78,7 @@ TEST_F(LogicalDnsClusterTest, Basic) {
 
   EXPECT_EQ(1UL, cluster_->hosts().size());
   EXPECT_EQ(1UL, cluster_->healthyHosts().size());
-  EXPECT_EQ(0UL, cluster_->localZoneHosts().size());
-  EXPECT_EQ(0UL, cluster_->localZoneHealthyHosts().size());
+  EXPECT_EQ(0UL, cluster_->healthyHostsPerZone().size());
```
Comment (mattklein123) [inline test/common/upstream/logical_dns_cluster_test.cc]: also EXPECT_EQ that hostsPerZone().size() == 0

File: include/envoy/upstream/upstream.h [hunk 2]
```diff
@@ -127,13 +127,20 @@ class HostSet {
    *         set on the command line and to use a cluster type that supports population such as
    *         the SDS cluster type.
    */
-  virtual const std::vector<HostPtr>& localZoneHosts() const PURE;
 
   /**
-   * @return all healthy hosts that are in the zone local to this node. See healthyHosts() and
-   *         localZoneHosts() for more information.
+   * @return hosts per zone, index 0 is dedicated to local zone hosts.
+   * If there is no hosts in local zone for upstream cluster hostPerZone() will @return
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: nit: "are no hosts"

File: test/common/upstream/logical_dns_cluster_test.cc (hunk 1 above)
Comment (RomanDzhabarov) [inline test/common/upstream/logical_dns_cluster_test.cc]: added

File: include/envoy/upstream/upstream.h (hunk 2 above)
Comment (RomanDzhabarov) [inline include/envoy/upstream/upstream.h]: fixed.

**Pull request merged**